test/transaction_test.exs

```diff
@@ -0,0 +1,78 @@
+defmodule TransactionTest do
+  use ExUnit.Case
+  import Mariaex.TestHelper
+
+  setup do
+    opts = [database: "mariaex_test", username: "mariaex_user", password: "mariaex_pass", backoff_type: :stop]
+    {:ok, pid} = Mariaex.Connection.start_link(opts)
+    {:ok, [pid: pid]}
+  end
+
+  test "transaction shows correct transaction status", context do
+    pid = context[:pid]
+    opts = [mode: :transaction]
+
+    assert DBConnection.status(pid, opts) == :idle
+    assert query("SELECT 42", []) == [[42]]
+    assert DBConnection.status(pid, opts) == :idle
+    {conn, _} = DBConnection.begin!(pid, opts)
+    assert DBConnection.status(conn, opts) == :transaction
+    DBConnection.commit!(conn, opts)
+    assert DBConnection.status(pid, opts) == :idle
+    assert query("SELECT 42", []) == [[42]]
+    assert DBConnection.status(pid) == :idle
+  end
+
+  test "can not begin transaction if already begun", context do
+    pid = context[:pid]
+    opts = [mode: :transaction]
+
+    {conn, _} = DBConnection.begin!(pid, opts)
+    assert {:error, %DBConnection.TransactionError{status: :transaction}} =
+      DBConnection.begin(conn, opts)
+  end
+
+  test "can not commit or rollback transaction if not begun", context do
+    pid = context[:pid]
+    opts = [mode: :transaction]
+
+    assert {:error, %DBConnection.TransactionError{status: :idle}} =
+      DBConnection.commit(pid, opts)
+    assert {:error, %DBConnection.TransactionError{status: :idle}} =
+      DBConnection.rollback(pid, opts)
+  end
+
+  test "savepoint transaction shows correct transaction status", context do
+    pid = context[:pid]
+    opts = [mode: :savepoint]
+
+    {conn, _} = DBConnection.begin!(pid, [mode: :transaction])
+    assert DBConnection.status(conn, opts) == :transaction
+
+    assert {:ok, conn, _} = DBConnection.begin(conn, opts)
+    assert DBConnection.status(conn, opts) == :transaction
+    DBConnection.commit!(conn, opts)
+    assert DBConnection.status(pid, opts) == :transaction
+
+    assert {:ok, conn, _} = DBConnection.begin(pid, opts)
+    assert DBConnection.status(conn, opts) == :transaction
+    DBConnection.rollback!(conn, opts)
+    assert DBConnection.status(pid, opts) == :transaction
+
+   DBConnection.commit!(pid, [mode: :transaction])
+   assert DBConnection.status(pid) == :idle
+   assert query("SELECT 42", []) == [[42]]
+  end
+
+  test "can not begin, commit or rollback savepoint transaction if not begun", context do
```

Comment:
we should mention savepoint in this name as its confusing to say can not begin if mnot not begun without that context.

Reply:
You're right, this test is wonderful ;)